abcix/example/kvstore/kvstore.go

```diff
@@ -0,0 +1,249 @@
+	"github.com/tendermint/tendermint/abcix/types"
+	"github.com/tendermint/tendermint/libs/log"
+	tdtypes "github.com/tendermint/tendermint/types"
+	"github.com/tendermint/tendermint/version"
+)
+
+var (
+	kvPairPrefixKey = []byte("kvPairKey:")
+
+	ProtocolVersion uint64 = 0x1
+	maxBytes               = tdtypes.DefaultConsensusParams().Block.MaxBytes
+	maxGas          int64  = 10
+)
+
+type State struct {
+	db            dbm.DB
+	Size          int64            `json:"size"`
+	Height        int64            `json:"height"`
+	ResourceUsage map[string]int64 `json:"resource_usage"`
+	AppHash       []byte           `json:"app_hash"`
+}
+
+func loadState(db dbm.DB) State {
+	return State{
+		db:            db,
+		ResourceUsage: make(map[string]int64),
+	}
+}
+
+func prefixKey(key []byte) []byte {
+	return append(kvPairPrefixKey, key...)
+}
+
+type Tx struct {
+	kvs []struct {
+		key []byte
+		val []byte
+	}
+	from     []byte
+	gasprice int64
+}
+
+//---------------------------------------------------
+
+var _ types.Application = (*Application)(nil)
+
+type Application struct {
+	types.BaseApplication
+	state        State
+	logger       log.Logger
+	RetainBlocks int64 // blocks to retain after commit (via ResponseCommit.RetainHeight)
+}
+
+func NewApplication() *Application {
+	state := loadState(dbm.NewMemDB())
+	return &Application{state: state, logger: log.NewTMLogger(log.NewSyncWriter(os.Stdout))}
+}
+
+func (app *Application) Info(req types.RequestInfo) (resInfo types.ResponseInfo) {
+	return types.ResponseInfo{
+		Data:             fmt.Sprintf("{\"size\":%v}", app.state.Size),
+		Version:          version.ABCIVersion,
+		AppVersion:       ProtocolVersion,
+		LastBlockHeight:  app.state.Height,
+		LastBlockAppHash: app.state.AppHash,
+	}
+}
+
+func (app *Application) CheckTx(req types.RequestCheckTx) types.ResponseCheckTx {
+	// Tx looks like "[key1]=[value1],[key2]=[value2],[from],[gasprice]"
+	// e.g. "a=41,c=42,alice,100"
+	tx, err := parseTx(req.Tx)
+	if err != nil {
+		app.logger.Error("failed to parse tx", "err", err)
+		return types.ResponseCheckTx{Code: code.CodeTypeEncodingError}
+	}
+	return types.ResponseCheckTx{Code: code.CodeTypeOK, GasWanted: int64(len(tx.kvs)), Priority: uint64(tx.gasprice)}
+}
+
+// Iterate txs from mempool ordered by priority and select the ones to be included in the next block
+func (app *Application) CreateBlock(
+	req types.RequestCreateBlock,
+	mempool *types.MempoolIter,
+) types.ResponseCreateBlock {
+	var txs, invalidTxs [][]byte
+	var size int64
+
+	remainBytes, remainGas := maxBytes, maxGas
+	for mempool.HasNext() {
+		tx, err := mempool.GetNextTransaction(remainBytes, remainGas)
+		if err != nil {
+			panic("failed to get next tx from mempool")
+		}
+		// Break if mempool runs out of tx
+		if len(tx) == 0 {
+			break
+		}
+
+		newState, gasUsed, err := executeTx(app.state, tx, true)
+		if err != nil {
+			invalidTxs = append(invalidTxs, tx)
+			continue
```

Comment:
does this mean get invalid tx from mempool will not affect `remainBytes` and `remainGas`?

Reply:
yup. they will be thrown away and not included in the block, thus no other nodes can verify how much gas they used